src/ngx_http_lua_worker_thread.c

```
@@ -0,0 +1,464 @@
+        lua_getfield(L, -1, "path");
+        path = lua_tolstring(L, -1, &path_len);
+        lua_getfield(L, -2, "cpath");
+        cpath = lua_tolstring(L, -1, &cpath_len);
+
+        lua_getglobal(vm, "package");
+        lua_pushlstring(vm, path, path_len);
+        lua_setfield(vm, -2, "path");
+        lua_pushlstring(vm, cpath, cpath_len);
+        lua_setfield(vm, -2, "cpath");
+
+        /* pop path and cpath from L */
+        lua_pop(L, 2);
+
+        /* pop the "package" table */
+        lua_pop(L, 1);
+        lua_pop(vm, 1);
+
+        ngx_thread_mutex_create(&ctx->mutex, ngx_cycle->log);
+
+    } else {
+        ctx = ctxpool->next;
+        ctxpool->next = ctxpool->next->next;
+    }
+
+    return ctx;
+}
+
+
+static void
+ngx_http_lua_free_task_ctx(ngx_http_lua_task_ctx_t *ctx)
+{
+    ctx->next = ctxpool->next;
+    ctxpool->next = ctx;
+    /*  clean Lua stack */
+    lua_settop(ctx->vm, 0);
+
+    worker_thread_vm_count--;
+}
+
+
+static int
+ngx_http_lua_xcopy(lua_State *from, lua_State *to, int idx,
+                         const int allow_nil)
+{
+    size_t           len = 0;
+    const char      *str;
+
+    switch (lua_type(from, idx)) {
+    case LUA_TBOOLEAN:
+        lua_pushboolean(to, lua_toboolean(from, idx));
+        return LUA_TBOOLEAN;
+
+    case LUA_TLIGHTUSERDATA:
+        lua_pushlightuserdata(to, lua_touserdata(from, idx));
+        return LUA_TLIGHTUSERDATA;
+
+    case LUA_TNUMBER:
+        lua_pushnumber(to, lua_tonumber(from, idx));
+        return LUA_TNUMBER;
+
+    case LUA_TSTRING:
+        str = lua_tolstring(from, idx, &len);
+        lua_pushlstring(to, str, len);
+        return LUA_TSTRING;
+
+    case LUA_TTABLE:
+        lua_newtable(to);
+        /* to positive number */
+        if (idx < 0) {
+            idx = lua_gettop(from) + idx + 1;
+        }
+
+        lua_pushnil(from);
+
+        while (lua_next(from, idx)) {
+            if (ngx_http_lua_xcopy(from, to, -2, 0) != LUA_TNONE) {
```

maybe throw an error when there are unsupported types of values inside the table is the better way?
ignore them silently may be difficult to debug for people who don't know this limitation well.